pyqtorch/noise.py

```diff
@@ -0,0 +1,356 @@
+from __future__ import annotations
+
+from math import sqrt
+
+import torch
+from torch import Tensor
+
+from pyqtorch.apply import operator_product
+from pyqtorch.matrices import DEFAULT_MATRIX_DTYPE, IMAT, XMAT, YMAT, ZMAT, _dagger
+from pyqtorch.utils import DensityMatrix, density_mat
+
+
+class Noise(torch.nn.Module):
+    def __init__(
+        self, kraus: list[Tensor], target: int, probabilities: tuple[float, ...] | float
+    ) -> None:
+        super().__init__()
+        self.target: int = target
+        self.qubit_support: tuple[int, ...] = (self.target,)
+        self.kraus: list = []
+        for index, tensor in enumerate(kraus):
+            self.register_buffer(f"kraus_{index}", tensor)
+            self.kraus.append(tensor)
+        self._device: torch.device = self.kraus[0].device
+        self.probabilities: tuple[float, ...] | float = probabilities
+
+    @property
+    def name(self) -> str:
+        return self.__class__.__name__
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, type(self)):
+            return self.name == other.name and self.probabilties == other.probabilties
+        return False
+
+    def extra_repr(self) -> str:
+        return f"'qubit_support'={self.qubit_support}, 'probabilities'={self.probabilities}"
+
+    @property
+    def kraus_operator(self) -> list[Tensor]:
```

Comment:
why the property? we already have the `kraus` field in the class, can we just give it the same name? so,  `self.kraus_operator` OR `self.kraus` ? but not have both

Reply:
You’re right. Having both the kraus field and the `kraus_operator` property is redundant. So I simplified the code to use a single kraus property that directly accesses the saved buffers.